tests/testmplex.nim

```diff
@@ -594,6 +594,468 @@ suite "Mplex":
 
     waitFor(testNewStream())
 
+  test "e2e - channel closes listener with EOF":
+    proc testNewStream() {.async.} =
+      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0").tryGet()
+
+      var listenStreams: seq[Connection]
+      proc connHandler(conn: Connection) {.async, gcsafe.} =
+        let mplexListen = Mplex.init(conn)
+        mplexListen.streamHandler = proc(stream: Connection)
+          {.async, gcsafe.} =
+          listenStreams.add(stream)
+          try:
+            discard await stream.readLp(1024)
+          except LPStreamEOFError:
+            await stream.close()
+            return
+
+          check false
+
+        await mplexListen.handle()
+        await mplexListen.close()
+
+      let transport1 = TcpTransport.init()
+      let listenFut = await transport1.listen(ma, connHandler)
+
+      let transport2: TcpTransport = TcpTransport.init()
+      let conn = await transport2.dial(transport1.ma)
+
+      let mplexDial = Mplex.init(conn)
+      let mplexDialFut = mplexDial.handle()
+      var dialStreams: seq[Connection]
+      for i in 0..9:
+        dialStreams.add((await mplexDial.newStream()))
+
+      for i, s in dialStreams:
+        await s.closeWithEOF()
+        await sleepAsync(10.millis)
```

Comment:
these sleeps should ideally not be needed - ie everything is deterministic in the test in the sense that if the close finishes, it should lead to ordering guarantees - if that's missing, it's likely an opportunity for a race condition

Reply:
yeah you right, I originally used `s.close()` which would fail to close because there where no readers/writers on the stream and the sleep is a leftover from that.